Initialized Firebase DB and Defined API Endpoints

rohanmaheshwari430 opened the pull request:
### Summary <!-- Required -->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Itemize bug fixes, new features, and other changes -->
<!-- Feel free to break this into sub-sections, i.e. features, fixes, etc. -->
<!-- Some examples are shown below. -->

This pull request creates an instance of the firebase firestore. It also contains the GET,POST, DELETE, and UPDATE API endpoints for the Courses page. In the process, I also made the Course type. Currently, I have structured it such that the client will pass the type of the course as the "id" and the code as a "code", and the the course content as its own JSON object.  After the completion of a request, the API will send a response containing a JSON object that has two or three fields, a success boolean, message string, and/or data that is essentially a confirmation of the request completion. 


Here is an example of the current POST request body format: 

![Screen Shot 2021-03-22 at 2 07 55 AM](https://user-images.githubusercontent.com/46412997/111947238-6c8ea980-8ab3-11eb-8bc8-a0270faa2966.png)

These changes were made in the server.ts file. 

### Test Plan 

To test these endpoints, cd into the server directory and type the command "yarn start". It may return an error saying you are missing the service account file, in which case please reach out to Jessica to get instructions on receiving a service account file. Make sure to rename it to the service account file name that the error mentions. If no error was returned, copy this Postman link at https://www.getpostman.com/collections/b2a30546c45e18202061 and import it as a collection in your personal Postman workspace. Now, you should be all set to test the endpoints. Each of these requests targets one of the four Courses-related API endpoints. Feel free to change the request body or query parameters to test different data. 

GET course (all courses)
<img width="866" alt="Screen Shot 2021-04-09 at 1 01 58 PM" src="https://user-images.githubusercontent.com/46412997/114215594-c5739400-9933-11eb-8870-1c9e051ad071.png">

GET courses (singular course)
<img width="865" alt="Screen Shot 2021-04-09 at 1 02 25 PM" src="https://user-images.githubusercontent.com/46412997/114215645-d58b7380-9933-11eb-8a38-b73dc7aa4c39.png">

CREATE course
<img width="874" alt="Screen Shot 2021-04-09 at 1 02 42 PM" src="https://user-images.githubusercontent.com/46412997/114215676-dfad7200-9933-11eb-8d96-b85c7274be48.png">

UPDATE course
<img width="868" alt="Screen Shot 2021-04-09 at 1 02 59 PM" src="https://user-images.githubusercontent.com/46412997/114215712-e9cf7080-9933-11eb-88df-6ddd50ac4b43.png">

DELETE course
<img width="872" alt="Screen Shot 2021-04-09 at 1 03 10 PM" src="https://user-images.githubusercontent.com/46412997/114215729-f05de800-9933-11eb-89d7-6f029584104a.png">


### Notes 

* **Jessica and I have restructured the DB architecture with the following format:**

    * Courses(collection)/Semester(document)/CourseType(collection)/CourseID(document)
 
* It may be worth considering to put the firebase and API endpoints in a different file in the future.

* All endpoints can be found at https://www.postman.com/collections/b2a30546c45e18202061

rohanmaheshwari430 committed:
Initialized Firebase DB and Defined API Endpoints

Initialized admin and created instance of firestore. Wrote the endpoints for get and post (Courses). Also created course data structure. Tested the endpoints via postman.

rohanmaheshwari430 committed:
Updates to GET and POST, Added in DELETE

*Implemented query parameter features in/getCourse to retrieve a specific course from the database.

*Implemented duplicate course check in /createCourse.

*Wrote /deleteCourse endpoint which also ensures the course to be deleted does exist.

jessfeng commented on server/server.ts:
```diff
@@ -1,11 +1,154 @@
+  "major": string,
+  "design_areas": string[],
+  "semester": string[]
+}
```
It might be cleaner if we put types in another file that specifies the structure of objects we use in our code.

jessfeng replied:
Oh you can probably disregard this comment, it was made before you made the change hehe

rohanmaheshwari430 committed:
Finished Update endpoint and modified other CRUD operations

Finished the Update endpoint and also went back to other endpoints to clean up code and reflect the new data structure format.

rohanmaheshwari430 committed:
Update server.ts

rohanmaheshwari430 committed:
restructured API to match new DB architecture

Changed every endpoints object structuring/destructuring to accomodate new DB design. All functionalities are the same.

jessfeng commented on server/server.ts:
```diff
@@ -1,11 +1,141 @@
+  "id": string, 
+  "code": number,
+  "content": course_content
+}
```
It might be cleaner to put all types/models in another models.ts file or models folder within the server folder, since we'll have a couple for each page on our site, but these look really nicely structured!

jessfeng reviewed (commented):
Congrats on your first big PR! This looks really nicely structured and has pretty neat code, one suggestion is to break off the models/types and put them in another file or folder, since it might clutter the code that's in server.ts, where we're writing endpoints. But other than that, it looks good!

rohanmaheshwari430 committed:
relocated types into a separate file

katrina433 reviewed (commented):
Looks very organized and structured! Just a question: for the course content type, I thought the course website is just the roster site, but apparently it's different. What does this site link to if the courses don't have their own sites? (Some courses use Canvas only and I don't think it's accessible by students who are not enrolled). But everything looks good!

rohanmaheshwari430 commented:
> Looks very organized and structured! Just a question: for the course content type, I thought the course website is just the roster site, but apparently it's different. What does this site link to if the courses don't have their own sites? (Some courses use Canvas only and I don't think it's accessible by students who are not enrolled). But everything looks good!

rohanmaheshwari430 commented:
> Looks very organized and structured! Just a question: for the course content type, I thought the course website is just the roster site, but apparently it's different. What does this site link to if the courses don't have their own sites? (Some courses use Canvas only and I don't think it's accessible by students who are not enrolled). But everything looks good!

Great question, I will talk to Jessica about this. You are right that many courses may not have their own site. My current understanding is that those fields will be null and there simply wont be a "course site" to show for. I will get back you on this.

rohanmaheshwari430 committed:
modified /getCourses

made it so that the endpoint returns the id and the code of the course in each course object.

katrina433 reviewed (approved):
Looks good!
